azure-iot-device/azure/iot/device/common/pipeline/pipeline_stages_base.py

```diff
@@ -303,17 +371,20 @@ def _block(self, op):
     def _unblock(self, op, error):
         """
         Unblock this stage after the connection is complete.  This also means
-        releasing all the queued up operations that we were waiting for the
-        connect operation to complete.
+        releasing all the operations that were queued up.
         """
         logger.info("{}({}): disabling block and releasing queued ops.".format(self.name, op.name))
         self.blocked = False
         logger.info(
             "{}({}): processing {} items in queue".format(self.name, op.name, self.queue.qsize())
         )
-        # loop through our queue and release all the blocked operations
-        while not self.queue.empty():
-            op_to_release = self.queue.get_nowait()
+        # Loop through our queue and release all the blocked operations
+        # Put a new Queue in self.queue because releasing ops might put them back in the
+        # queue, especially if there's a ConnectOperation in the list of ops to release
+        old_queue = self.queue
+        self.queue = queue.Queue()
+        while not old_queue.empty():
```

Comment:
>while not old_queue.empty(): [](start = 8, length = 28)

What happens if new ops are added to the new queue while we're still resolving the ops old queue? Or does the threading model protect us from this? It would seem to me that the ops in the old queue should ALL be resolved before anything in the new queue, since they predate it.

e.g. 5 Publish ops are queued while waiting for a ConnectOp to resolve. The connect is resolved, and we begin resolving the 5 publishes from the old_queue, while a new empty queue is made. While we're resolving Publish #2, a DisconnectOp is sent down the pipe. The stage is unblocked already, since it's that happens before the resolution of all the old queued ops, so does the DisconnectOp block on our remaining Publishes or not?

If it does, cool. If it doesn't, then what happens to these Publish ops that are caught in limbo? And what happens if a ConnectOp is queued immediately after the DisconnectOp?

Reply:
The answer to this depends on what you mean by "resolved."  That is not a term that we've used here.  We can only ensure that ops get passed down to the next lower level.  We never wait for ops to complete (e.g. ACK comes back) unless we absolutely must (such as waiting for a CONACK before publishing).

Here are a few examples that I hope are self-explanatory.

[http://www.plantuml.com/plantuml/svg/0/pLPDRy8m3Br7wd_maKdZG7P3wiHbFoIafeqMqgf91zkldzFIxr8oYSD0Wg8ysy_FBnO7wqJfghEEM1p3nrMS2usmsuCe2f2vCIYT1M4IAAgZLZP32noSEjFt9LE4lGZbb4dXb9Tj0BYS7dqcXn374QkA0alOleF5KWcjVX3sK5TkGoBM7otYZZgNdrGdOYQdhALACxS7hKxeoJJ7a9j19CMb2CeaIWfNqqK3BXEkFR5D52OH4zAfBqBruHD2EhzG1pFIAutTzDhkWb0yOas92QfLr5PI8YQUwrsXAbFt4x44_zZFW6i0YazGJNkBNBkUManNAGZ5XwYHGpcH4ZHtONlJlW9U9HRTUE0abBQr0PzimNzlbxhp9rbbFhgbxFFHhQ4cF4UZVLiSRKONNljBN--QoX1hkkrIQDql0J_fjCohejbUT6RRIfkNY06zGcpWSS5MyE0b2MT9xbW2JJ8dVHDnqqd-cFt0Fq4fUNXAFahPzA75DtsE9Av_7OLtbKySkivMrUvywb7BNdqWz42NG2HVGAwik5Lf5jVTI9tLnHYKcmVdvhFS_dsW9_e3y0i0](http://www.plantuml.com/plantuml/svg/0/pLPDRy8m3Br7wd_maKdZG7P3wiHbFoIafeqMqgf91zkldzFIxr8oYSD0Wg8ysy_FBnO7wqJfghEEM1p3nrMS2usmsuCe2f2vCIYT1M4IAAgZLZP32noSEjFt9LE4lGZbb4dXb9Tj0BYS7dqcXn374QkA0alOleF5KWcjVX3sK5TkGoBM7otYZZgNdrGdOYQdhALACxS7hKxeoJJ7a9j19CMb2CeaIWfNqqK3BXEkFR5D52OH4zAfBqBruHD2EhzG1pFIAutTzDhkWb0yOas92QfLr5PI8YQUwrsXAbFt4x44_zZFW6i0YazGJNkBNBkUManNAGZ5XwYHGpcH4ZHtONlJlW9U9HRTUE0abBQr0PzimNzlbxhp9rbbFhgbxFFHhQ4cF4UZVLiSRKONNljBN--QoX1hkkrIQDql0J_fjCohejbUT6RRIfkNY06zGcpWSS5MyE0b2MT9xbW2JJ8dVHDnqqd-cFt0Fq4fUNXAFahPzA75DtsE9Av_7OLtbKySkivMrUvywb7BNdqWz42NG2HVGAwik5Lf5jVTI9tLnHYKcmVdvhFS_dsW9_e3y0i0)


[http://www.plantuml.com/plantuml/svg/0/xLV1ReCm3BqZyG_MTcaF75PjErHgLKtQ_YCbRecQ9eY4iUthvy1e8fgoxZ2fqmedmB6Vd_CUi3AMLRRUopHPBE3vZUrBYN0_XrmhXRdDbcMzbi8KsP8BCxm94wivUuSi0oOb-0W3J76eq5zM21jTvSYdi5YaILsMM46s18ELO59y8CoXnT9NJXF_A1gtbZhV2RLD4vPRySei-g31OgaRgdPKO0H4t_NTQ9Hl-AnaipJfa0oYldA0gVCSaPiqKPgwgSIsi7EGOeDMxB5h6BI2cpRB3GZ5HKvz6sWAL60hfaofgQ0mVKxa0JtjiX5UQzMMI1EEFwGqeEkRgOMK_TQ0mnfHmdr0pxdZAUZwzvcQ3QAwxNCAsg6BFkn0E6oOaACXiqE8EmELFvojygPW5XhQaKWoVuU6AUiqhxJAEb67qhyA_0o1_mDDXeiR9gpp__OrKIBvbjg2BWrOJUIutMWnbX9kNrwUw0WTCKs03kFUBHmcfL2xQSGZVGe2liNMAvrLTgixltIx-kkfQI4wL0xbIOHq7noI4tFsgepRIgIlg0b4yrmDR-XcWPSzKZb6ysV6snkiQt8jOGl_8NArfZzcJOV9NQupNPGJ17mTgpkAr6iZg701lX35DM8Bexwsei7IJyud](http://www.plantuml.com/plantuml/svg/0/xLV1ReCm3BqZyG_MTcaF75PjErHgLKtQ_YCbRecQ9eY4iUthvy1e8fgoxZ2fqmedmB6Vd_CUi3AMLRRUopHPBE3vZUrBYN0_XrmhXRdDbcMzbi8KsP8BCxm94wivUuSi0oOb-0W3J76eq5zM21jTvSYdi5YaILsMM46s18ELO59y8CoXnT9NJXF_A1gtbZhV2RLD4vPRySei-g31OgaRgdPKO0H4t_NTQ9Hl-AnaipJfa0oYldA0gVCSaPiqKPgwgSIsi7EGOeDMxB5h6BI2cpRB3GZ5HKvz6sWAL60hfaofgQ0mVKxa0JtjiX5UQzMMI1EEFwGqeEkRgOMK_TQ0mnfHmdr0pxdZAUZwzvcQ3QAwxNCAsg6BFkn0E6oOaACXiqE8EmELFvojygPW5XhQaKWoVuU6AUiqhxJAEb67qhyA_0o1_mDDXeiR9gpp__OrKIBvbjg2BWrOJUIutMWnbX9kNrwUw0WTCKs03kFUBHmcfL2xQSGZVGe2liNMAvrLTgixltIx-kkfQI4wL0xbIOHq7noI4tFsgepRIgIlg0b4yrmDR-XcWPSzKZb6ysV6snkiQt8jOGl_8NArfZzcJOV9NQupNPGJ17mTgpkAr6iZg701lX35DM8Bexwsei7IJyud)



---
In reply to: [311714638](https://github.com/Azure/azure-iot-sdk-python-preview/pull/160#discussion_r311714638) [](ancestors = 311714638)

Reply:
And, for your last question about calling disconnect then immediately calling connect:

[http://www.plantuml.com/plantuml/svg/0/lPFBRi8m44NtaV8FiwGBraX8GTdm8OCz0ek9RMAxBVrw1imZXK1JXCZ2Y-Z8z_gCNOU8RKmDbqLLmVmR6yy4aocWzw2GEO0sGJbhIKL0go6ko89f6j868_7sI9sGgYgBPUkIxzrXz54M85_odbeOpo1GQv3D3y4K3bKwS4OlH2--OAVMnYupZ2gQJwbr0NisOVSbuNrvPFLj--7LX_K-ozIQHAU3JQ94ly9pGyQjh3T3sFzOe5hdc8nUHNJea9GYagG_MYLxC9afJOD4FclWHPEx7vPdTC_3Gz7_iFmEnQUJ7BFambmVf9PLtkyE](http://www.plantuml.com/plantuml/svg/0/lPFBRi8m44NtaV8FiwGBraX8GTdm8OCz0ek9RMAxBVrw1imZXK1JXCZ2Y-Z8z_gCNOU8RKmDbqLLmVmR6yy4aocWzw2GEO0sGJbhIKL0go6ko89f6j868_7sI9sGgYgBPUkIxzrXz54M85_odbeOpo1GQv3D3y4K3bKwS4OlH2--OAVMnYupZ2gQJwbr0NisOVSbuNrvPFLj--7LX_K-ozIQHAU3JQ94ly9pGyQjh3T3sFzOe5hdc8nUHNJea9GYagG_MYLxC9afJOD4FclWHPEx7vPdTC_3Gz7_iFmEnQUJ7BFambmVf9PLtkyE)


---
In reply to: [312164161](https://github.com/Azure/azure-iot-sdk-python-preview/pull/160#discussion_r312164161) [](ancestors = 312164161,311714638)